src/tellers/IntentsTeller.sol

```diff
@@ -0,0 +1,630 @@
+        );
+
+        Asset memory asset = _beforeDeposit(depositData.asset);
+        shares = asset.sharePremium > 0 ? shares.mulDivDown(1e4 - asset.sharePremium, 1e4) : shares;
+        if (shares < depositData.minimumOut) {
+            revert IntentsTeller__MinimumMintNotMet();
+        }
+
+        vault.enter(depositData.user, depositData.asset, depositData.amountIn, depositData.to, shares);
+
+        if (enforceShareLock) {
+            _afterPublicDeposit(depositData.user, depositData.asset, depositData.amountIn, shares, shareLockPeriod);
+        } else {
+            emit BulkDeposit(address(depositData.asset), depositData.amountIn);
+        }
+    }
+
+    function _erc20Withdraw(ActionData memory withdrawData) internal returns (uint256 assetsOut) {
+        Asset memory asset = assetData[withdrawData.asset];
+        if (!asset.allowWithdraws) {
+            revert IntentsTeller__AssetNotSupported();
+        }
+
+        if (withdrawData.amountIn == 0) revert IntentsTeller__ZeroShares();
+
+        _verifySignedMessage(withdrawData);
+
+        fluxManager.refreshInternalFluxAccounting();
+
+        assetsOut = withdrawData.amountIn.mulDivDown(
+            fluxManager.getRateSafe(withdrawData.rate, withdrawData.asset == token0), ONE_SHARE
+        ); // check rate direction
+
+        if (assetsOut < withdrawData.minimumOut) {
+            revert IntentsTeller__MinimumAssetsNotMet();
+        }
+
+        vault.exit(withdrawData.to, withdrawData.asset, assetsOut, withdrawData.user, withdrawData.amountIn);
+        emit Withdraw(address(withdrawData.asset), withdrawData.amountIn);
+    }
+
+    /**
+     * @notice Handle pre-deposit checks.
+     */
+    function _beforeDeposit(ERC20 depositAsset) internal view returns (Asset memory asset) {
+        asset = assetData[depositAsset];
+        if (!asset.allowDeposits) {
+            revert IntentsTeller__AssetNotSupported();
+        }
+    }
+
+    /**
+     * @notice Handle share lock logic, and event.
+     */
+    function _afterPublicDeposit(
+        address user,
+        ERC20 depositAsset,
+        uint256 depositAmount,
+        uint256 shares,
+        uint256 currentShareLockPeriod
+    ) internal {
+        // Increment then assign as its slightly more gas efficient.
+        uint256 nonce = ++depositNonce;
+        // Only set share unlock time and history if share lock period is greater than 0.
+        if (currentShareLockPeriod > 0) {
+            beforeTransferData[user].shareUnlockTime = block.timestamp + currentShareLockPeriod;
+            publicDepositHistory[nonce] = keccak256(
+                abi.encode(user, depositAsset, depositAmount, shares, block.timestamp, currentShareLockPeriod)
+            );
+        }
+        emit Deposit(nonce, user, address(depositAsset), depositAmount, shares, block.timestamp, currentShareLockPeriod);
+    }
+
+    function _verifySignedMessage(ActionData memory actionData) internal {
```

Comment:
I think you can make any `ActionData` function arguments use `calldata` instead of `memory` to save some gas.